File: Assets/Game/Scripts/UI/WinScoreUIControl.lua
local Yield = UnityEngine.Yield
local WaitForSeconds = UnityEngine.WaitForSeconds
local KeyCode = UnityEngine.KeyCode
local Text = UnityEngine.UI.Text
local GameSoundType = Ballance2.Services.GameSoundType
local GameUIManager = Ballance2.Services.GameManager.GetSystemService('GameUIManager') ---@type GameUIManager
local I18N = Ballance2.Services.I18N.I18N

---过关之后的分数统计界面控制
---@class WinScoreUIControl : GameLuaObjectHostClass
---@field ScoreTotal Text
---@field ScoreExtraLives Text
---@field ScoreBouns Text
---@field ScoreTimePoints Text
---@field HighlightBar1 GameObject
---@field HighlightBar2 GameObject
---@field HighlightBar3 GameObject
---@field HighlightBar4 GameObject
---@field _GamePlayManager GamePlayManager
---@field _CountSound AudioSource
---@field _HighscoreSound AudioSource
---@field _SwitchSound AudioSource
---@field _CountingPointEndCallback function
---@field _IsInSeq boolean
---@field _ScoreNTotal number
WinScoreUIControl = ClassicObject:extend()

function WinScoreUIControl:new() 
  self._ThisTimeHasNewHighscore = false
  self._CountingPointEndCallback = nil ---@type function
  GameUI.WinScoreUIControl = self
end
function WinScoreUIControl:Start() 
  self._SwitchSound = Game.SoundManager:RegisterSoundPlayer(GameSoundType.UI, "core.sounds:Menu_dong.wav", false, true, 'WinScoreUISwitch')
  self._HighscoreSound = Game.SoundManager:RegisterSoundPlayer(GameSoundType.UI, "core.sounds.music:Music_Highscore.wav", false, true, 'WinScoreUISwitch')
  self._CountSound = Game.SoundManager:RegisterSoundPlayer(GameSoundType.UI, "core.sounds:Menu_counter.wav", false, true, 'WinScoreUICount')
  GameUI.WinScoreUIControl = self
end
function WinScoreUIControl:Update() 
  if self._IsCountingPoint then
    if self._GamePlayManager.CurrentPoint > 0 then
      if self._GamePlayManager.CurrentPoint > 4000 then
        self._GamePlayManager.CurrentPoint = self._GamePlayManager.CurrentPoint - 20
        self._ScoreNTimePoints = self._ScoreNTimePoints + 20
        self._ScoreNTotal = self._ScoreNTotal + 20
      elseif self._GamePlayManager.CurrentPoint > 2000 then
        self._GamePlayManager.CurrentPoint = self._GamePlayManager.CurrentPoint - 10
        self._ScoreNTimePoints = self._ScoreNTimePoints + 10
        self._ScoreNTotal = self._ScoreNTotal + 10
      elseif self._GamePlayManager.CurrentPoint > 2000 then
        self._GamePlayManager.CurrentPoint = self._GamePlayManager.CurrentPoint - 10
        self._ScoreNTimePoints = self._ScoreNTimePoints + 10
        self._ScoreNTotal = self._ScoreNTotal + 10
      elseif self._GamePlayManager.CurrentPoint > 4 then
        self._GamePlayManager.CurrentPoint = self._GamePlayManager.CurrentPoint - 4
        self._ScoreNTimePoints = self._ScoreNTimePoints + 4
        self._ScoreNTotal = self._ScoreNTotal + 4
      else
        self._ScoreNTimePoints = self._ScoreNTimePoints + self._GamePlayManager.CurrentPoint
        self._ScoreNTotal = self._ScoreNTotal + self._GamePlayManager.CurrentPoint
        self._GamePlayManager.CurrentPoint = 0
      end
      
      GameUI.GamePlayUI:SetPointText(self._GamePlayManager.CurrentPoint)
      self._CountSound:Stop()
      self._CountSound:Play()
      self.ScoreTimePoints.text = tostring(self._ScoreNTimePoints)
      self.ScoreTotal.text = tostring(self._ScoreNTotal)
    else
      self._IsCountingPoint = false
      if type(self._CountingPointEndCallback) == "function" then
        self._CountingPointEndCallback()
      end
    end
  end
end

---开始分数统计序列
function WinScoreUIControl:StartSeq() 
  self._GamePlayManager = GamePlay.GamePlayManager
  self._IsInSeq = true
  self._ScoreNExtraLives = 0
  self._ScoreNTimePoints = 0
  self._ScoreNTotal = 0
  self._Skip = false
  self.HighlightBar1:SetActive(false)
  self.HighlightBar2:SetActive(false)
  self.HighlightBar3:SetActive(false)
  self.HighlightBar4:SetActive(false)

  coroutine.resume(coroutine.create(function()

    Yield(WaitForSeconds(5))
    
    self.EscKeyID = GameUIManager:WaitKey(KeyCode.Escape, false, function () self:Skip() end)
    self.ReturnKeyID = GameUIManager:WaitKey(KeyCode.Return, false, function () self:Skip() end)

    Game.UIManager:GoPage('PageEndScore')

    Yield(WaitForSeconds(1.7))
    if self._Skip then return end
    
    --关卡分数
    self._SwitchSound:Play()
    self.HighlightBar1:SetActive(true)
    self.ScoreBouns.text = tostring(self._GamePlayManager.LevelScore)
    self._ScoreNTotal = self._GamePlayManager.LevelScore
    self.ScoreTotal.text = tostring(self._ScoreNTotal)

    Yield(WaitForSeconds(1.7))
    if self._Skip then return end

    --额外时间点

    self._SwitchSound:Play()
    self.HighlightBar1:SetActive(false)
    self.HighlightBar2:SetActive(true)
    self._IsCountingPoint = true
    
  end))

  self._CountingPointEndCallback = function ()   
    coroutine.resume(coroutine.create(function()
      if self._Skip then return end

      Yield(WaitForSeconds(1.5))

      --额外生命点
      
      self._SwitchSound:Play()
      self.HighlightBar2:SetActive(false)
      self.HighlightBar3:SetActive(true)

      for i = self._GamePlayManager.CurrentLife, 1, -1 do

        Yield(WaitForSeconds(0.6))
        if self._Skip then return end

        GameUI.GamePlayUI:RemoveLifeBall()
        self._ScoreNExtraLives = self._ScoreNExtraLives + 200
        self._ScoreNTotal = self._ScoreNTotal + 200
        self.ScoreExtraLives.text = tostring(self._ScoreNExtraLives)
        self.ScoreTotal.text = tostring(self._ScoreNTotal)

      end

      Yield(WaitForSeconds(1.5))

      --完整分数
      if self._Skip then return end

      self._SwitchSound:Play()
      self.HighlightBar3:SetActive(false)
      self.HighlightBar4:SetActive(true)

      Yield(WaitForSeconds(5))
      if self._Skip then return end

      self._IsInSeq = false
      self:_ShowHighscore()
    end))
  end
end
function WinScoreUIControl:IsInSeq() return self._IsInSeq end
---跳过分数统计序列
function WinScoreUIControl:Skip() 
  if self._Skip then return end
  
  self._IsInSeq = false
  self._Skip = true
  self._IsCountingPoint = false
  self._SwitchSound:Play()
  self.HighlightBar1:SetActive(false)
  self.HighlightBar2:SetActive(false)
  self.HighlightBar3:SetActive(false)
  self.HighlightBar4:SetActive(true)

  local GamePlayManager = self._GamePlayManager

  self._ScoreNTimePoints = self._ScoreNTimePoints + GamePlayManager.CurrentPoint
  self._ScoreNExtraLives = self._ScoreNExtraLives + 200 * GamePlayManager.CurrentLife
  self._ScoreNTotal = GamePlayManager.LevelScore + self._ScoreNTimePoints + self._ScoreNExtraLives

  GamePlayManager.CurrentLife = 0
  GamePlayManager.CurrentPoint = 0
  GameUI.GamePlayUI:SetPointText(GamePlayManager.CurrentPoint)
  GameUI.GamePlayUI:SetLifeBallCount(0)
  
  self.ScoreTimePoints.text = tostring(self._ScoreNTimePoints)
  self.ScoreBouns.text = tostring(GamePlayManager.LevelScore)
  self.ScoreExtraLives.text = tostring(self._ScoreNExtraLives)
  self.ScoreTotal.text = tostring(self._ScoreNTotal)

  LuaTimer.Add(3500, function() self:_ShowHighscore() end)
end
function WinScoreUIControl:SaveHighscore(entryName) 
  Game.HighScoreManager.AddItem(GamePlay.GamePlayManager.CurrentLevelName, entryName, self._ScoreNTotal)
  Game.HighScoreManager.AddLevelPassState(GamePlay.GamePlayManager.NextLevelName)
end
function WinScoreUIControl:_ShowHighscore() 
  if self.EscKeyID ~= nil then
    GameUIManager:DeleteKeyListen(self.EscKeyID)
    self.EscKeyID = nil
  end
  if self.ReturnKeyID ~= nil then
    GameUIManager:DeleteKeyListen(self.ReturnKeyID)
    self.ReturnKeyID = nil
  end

  Game.UIManager:GoPage('PageHighscoreEntry')

  --重置文字为0
  self.ScoreTimePoints.text = '0'
  self.ScoreBouns.text = '0'
  self.ScoreExtraLives.text = '0'
  self.ScoreTotal.text = '0'

  --检查是不是新的高分
  local PageHighscoreEntry = Game.UIManager:GetCurrentPage()
  local HighscoreEntryNameTextScore = PageHighscoreEntry.Content:Find('TextScore'):GetComponent(Text) ---@type Text
  
  HighscoreEntryNameTextScore.text = tostring(self._ScoreNTotal)..' <size=20>'..I18N.Tr('core.ui.WinUIPoints')..'</size>'
  
  self._HighscoreSound:Play()
  if Game.HighScoreManager.CheckLevelHighScore(GamePlay.GamePlayManager.CurrentLevelName, self._ScoreNTotal) then
    PageHighscoreEntry.Content:Find('TextNewHighScore').gameObject:SetActive(true)
    PageHighscoreEntry.Content:Find('TextWin').gameObject:SetActive(false)
    self._ThisTimeHasNewHighscore = true
  else
    PageHighscoreEntry.Content:Find('TextNewHighScore').gameObject:SetActive(false)
    PageHighscoreEntry.Content:Find('TextWin').gameObject:SetActive(true)
    self._ThisTimeHasNewHighscore = false
  end
end

function CreateClass:WinScoreUIControl() 
  return WinScoreUIControl
end

File: Assets/Game/Menu/Scripts/Game/WinScoreUIControl.cs
using Ballance2.Base;
using Ballance2.Game.GamePlay;
using Ballance2.Services;
using Ballance2.Services.I18N;
using System.Collections;
using TMPro;
using UnityEngine;
using UnityEngine.UI;

namespace Ballance2.Menu
{
  public class WinScoreUIControl : GameSingletonBehavior<WinScoreUIControl> {
    public TMP_Text ScoreTotal;
    public TMP_Text ScoreExtraLives;
    public TMP_Text ScoreBouns;
    public TMP_Text ScoreTimePoints;
    public GameObject HighlightBar1;
    public GameObject HighlightBar2;
    public GameObject HighlightBar3;
    public GameObject HighlightBar4;
    public GameObject Panel;

    private AudioSource _CountSound;
    private AudioSource _HighscoreSound;
    private AudioSource _SwitchSound;
    private bool _IsInSeq = false;
    private bool _Skip = false;
    private bool _IsCountingPoint = false;
    private int _ScoreNTotal = 0;
    private int _ScoreNTimePoints = 0;
    private int _ScoreNExtraLives = 0;
    private GamePlayManager _GamePlayManager;
    private int EscKeyID = 0;
    private int ReturnKeyID = 0;

    public bool ThisTimeHasNewHighscore { get; private set; }

    private void Start() {
      GameTimer.Delay(2.4f, () => {
        var SoundManager = GameSoundManager.Instance;
        this._SwitchSound = SoundManager.RegisterSoundPlayer(GameSoundType.UI, "core.sounds:Menu_dong.wav", false, true, "WinScoreUISwitch");
        this._HighscoreSound = SoundManager.RegisterSoundPlayer(GameSoundType.UI, "core.sounds.music:Music_Highscore.wav", false, true, "WinScoreUISwitch");
        this._CountSound = SoundManager.RegisterSoundPlayer(GameSoundType.UI, "core.sounds:Menu_counter.wav", false, true, "WinScoreUICount");
      });
    }
    private void Update() {
      if (_IsCountingPoint) {
        if (_GamePlayManager.CurrentPoint > 0) {
          if (_GamePlayManager.CurrentPoint > 4000) {
            _GamePlayManager.CurrentPoint -= 20;
            _ScoreNTimePoints += 20;
            _ScoreNTotal += 20;
          }
          else if (_GamePlayManager.CurrentPoint > 2000) {
            _GamePlayManager.CurrentPoint -= 10;
            _ScoreNTimePoints += 10;
            _ScoreNTotal += 10;
          }
          else if (_GamePlayManager.CurrentPoint > 2000) {
            _GamePlayManager.CurrentPoint -= 10;
            _ScoreNTimePoints += 10;
            _ScoreNTotal += 10;
          }
          else if (_GamePlayManager.CurrentPoint > 4) {
            _GamePlayManager.CurrentPoint -= 4;
            _ScoreNTimePoints += 4;
            _ScoreNTotal += 4;
          }
          else {
            _ScoreNTimePoints += _GamePlayManager.CurrentPoint;
            _ScoreNTotal += _GamePlayManager.CurrentPoint;
            _GamePlayManager.CurrentPoint = 0;
          }
          
          GamePlayUIControl.Instance.SetPointText(_GamePlayManager.CurrentPoint);
          _CountSound.Stop();
          _CountSound.Play();
          ScoreTimePoints.text = _ScoreNTimePoints.ToString();
          ScoreTotal.text = _ScoreNTotal.ToString();
        }
        else {
          _IsCountingPoint = false;
          StartCoroutine(CountingPointEnd());
        }
      }
    }
    protected override void OnDestroy() 
    {
      Destroy(this._SwitchSound);
      Destroy(this._HighscoreSound);
      Destroy(this._CountSound);
    }

    /// <summary>
    /// 开始分数统计序列
    /// </summary>
    public void StartSeq() {
      _IsInSeq = true;
      _ScoreNExtraLives = 0;
      _ScoreNTimePoints = 0;
      _ScoreNTotal = 0;
      _Skip = false;
      _GamePlayManager = GamePlayManager.Instance;
      HighlightBar1.SetActive(false);
      HighlightBar2.SetActive(false);
      HighlightBar3.SetActive(false);
      HighlightBar4.SetActive(false);

      Panel.SetActive(false);
      GameUIManager.Instance.GoPage("PageEndScore");

      StartCoroutine(StartSeqInner());
    }
    private IEnumerator StartSeqInner() {
      yield return new WaitForSeconds(5);

      Panel.SetActive(true);
    
      EscKeyID = GameUIManager.Instance.WaitKey(KeyCode.Escape, false, Skip);
      ReturnKeyID = GameUIManager.Instance.WaitKey(KeyCode.Return, false, Skip);

      yield return new WaitForSeconds(1.7f);
      if (_Skip) 
        yield break;
      
      //关卡分数
      _SwitchSound.Play();
      HighlightBar1.SetActive(true);
      ScoreBouns.text = _GamePlayManager.LevelScore.ToString();
      _ScoreNTotal = _GamePlayManager.LevelScore;
      ScoreTotal.text = _ScoreNTotal.ToString();

      yield return new WaitForSeconds(1.7f);
      if (_Skip) 
        yield break;

      //额外时间点
      _SwitchSound.Play();
      HighlightBar1.SetActive(false);
      HighlightBar2.SetActive(true);
      _IsCountingPoint = true;
    }
    private IEnumerator CountingPointEnd() {
      if (_Skip) 
        yield break;

      yield return new WaitForSeconds(1.5f);

      var GamePlayUI = GamePlayUIControl.Instance;

      //额外生命点
      _SwitchSound.Play();
      HighlightBar2.SetActive(false);
      HighlightBar3.SetActive(true);

      for (var i = _GamePlayManager.CurrentLife; i > 0; i--)
      {
        yield return new WaitForSeconds(0.6f);
        if (_Skip) 
          yield break;

        GamePlayUI.RemoveLifeBall();
        _ScoreNExtraLives += 200;
        _ScoreNTotal += 200;
        ScoreExtraLives.text = _ScoreNExtraLives.ToString();
        ScoreTotal.text = _ScoreNTotal.ToString();
      }

      yield return new WaitForSeconds(1.5f);

      //完整分数
      if (_Skip) 
        yield break;

      _SwitchSound.Play();
      HighlightBar3.SetActive(false);
      HighlightBar4.SetActive(true);

      yield return new WaitForSeconds(5);
      if (_Skip) 
        yield break;

      _IsInSeq = false;
      _ShowHighscore();
    }

    /// <summary>
    /// 获取是否在分数统计序列中
    /// </summary>
    public void Skip() {
      if (_Skip)
        return;
  
      _IsInSeq = false;
      _Skip = true;
      _IsCountingPoint = false;
      _SwitchSound.Play();
      HighlightBar1.SetActive(false);
      HighlightBar2.SetActive(false);
      HighlightBar3.SetActive(false);
      HighlightBar4.SetActive(true);

      _ScoreNTimePoints += _GamePlayManager.CurrentPoint;
      _ScoreNExtraLives = _ScoreNExtraLives + 200 * _GamePlayManager.CurrentLife;
      _ScoreNTotal = _GamePlayManager.LevelScore + _ScoreNTimePoints + _ScoreNExtraLives;

      _GamePlayManager.CurrentLife = 0;
      _GamePlayManager.CurrentPoint = 0;
      GamePlayUIControl.Instance.SetPointText(_GamePlayManager.CurrentPoint);
      GamePlayUIControl.Instance.SetLifeBallCount(0);
      
      ScoreTimePoints.text = _ScoreNTimePoints.ToString();
      ScoreBouns.text = _GamePlayManager.LevelScore.ToString();
      ScoreExtraLives.text = _ScoreNExtraLives.ToString();
      ScoreTotal.text = _ScoreNTotal.ToString();

      GameTimer.Delay(3.5f, _ShowHighscore);
    }    
    /// <summary>
    /// 跳过分数统计序列
    /// </summary>
    /// <returns></returns>
    public bool IsInSeq() { return _IsInSeq; }

    //保存高分数据
    public void SaveHighscore(string entryName) {
      HighscoreManager.Instance.AddItem(_GamePlayManager.CurrentLevelName, entryName, _ScoreNTotal);
      HighscoreManager.Instance.AddLevelPassState(_GamePlayManager.NextLevelName);
    }

    //按ESC键直接进入高分界面
    private void _ShowHighscore() {
      if (EscKeyID != 0) 
      {
        GameUIManager.Instance.DeleteKeyListen(EscKeyID);
        EscKeyID = 0;
      }
      if (ReturnKeyID != 0) {
        GameUIManager.Instance.DeleteKeyListen(ReturnKeyID);
        ReturnKeyID = 0;
      }

      GameUIManager.Instance.GoPage("PageHighscoreEntry");

      //重置文字为0
      ScoreTimePoints.text = "0";
      ScoreBouns.text = "0";
      ScoreExtraLives.text = "0";
      ScoreTotal.text = "0";

      //检查是不是新的高分
      var PageHighscoreEntry = GameUIManager.Instance.GetCurrentPage();
      var HighscoreEntryNameTextScore = PageHighscoreEntry.Content.Find("TextScore").GetComponent<TMP_Text>();
      
      HighscoreEntryNameTextScore.text = $"{_ScoreNTotal} <size=20>{I18N.Tr("core.ui.WinUIPoints")}</size>";
      
      _HighscoreSound.Play();
      if(HighscoreManager.Instance.CheckLevelHighScore(_GamePlayManager.CurrentLevelName, _ScoreNTotal))
      {
        PageHighscoreEntry.Content.Find("TextNewHighScore").gameObject.SetActive(true);
        PageHighscoreEntry.Content.Find("TextWin").gameObject.SetActive(false);
        ThisTimeHasNewHighscore = true;
      }
      else
      {
        PageHighscoreEntry.Content.Find("TextNewHighScore").gameObject.SetActive(false);
        PageHighscoreEntry.Content.Find("TextWin").gameObject.SetActive(true);
        ThisTimeHasNewHighscore = false;
      }
    }
  }
}